File: backend/app/models/item.py
from datetime import datetime
from enum import Enum

from sqlalchemy import DateTime, Enum as SqlEnum, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class ItemStatus(str, Enum):
    LOST = "lost"
    FOUND = "found"


class ItemLifecycle(str, Enum):
    ACTIVE = "active"
    RESOLVED = "resolved"
    DELETED = "deleted"


class Item(Base):
    __tablename__ = "items"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    title: Mapped[str] = mapped_column(String(120), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    category: Mapped[str] = mapped_column(String(60), nullable=False, default="Other")
    location: Mapped[str] = mapped_column(String(120), nullable=False)
    status: Mapped[ItemStatus] = mapped_column(
        SqlEnum(
            ItemStatus,
            name="item_status",
            values_callable=lambda enum_cls: [item.value for item in enum_cls],
            validate_strings=True,
        ),
        nullable=False,
    )
    lifecycle: Mapped[ItemLifecycle] = mapped_column(
        SqlEnum(
            ItemLifecycle,
            name="item_lifecycle",
            values_callable=lambda enum_cls: [item.value for item in enum_cls],
            validate_strings=True,
        ),
        nullable=False,
        default=ItemLifecycle.ACTIVE,
    )
    contact_name: Mapped[str] = mapped_column(String(80), nullable=False)
    telegram_username: Mapped[str | None] = mapped_column(String(80), nullable=True)
    telegram_user_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

File: frontend/src/types/item.ts
export type ItemStatus = 'lost' | 'found'
export type ItemLifecycle = 'active' | 'resolved' | 'deleted'
export type ModerationStatus = 'pending' | 'approved' | 'rejected' | 'flagged'

export interface Item {
  id: number
  title: string
  description: string
  category: string
  location: string
  status: ItemStatus
  lifecycle: ItemLifecycle
  moderation_status: ModerationStatus
  moderation_reason?: string | null
  moderated_at?: string | null
  moderated_by?: string | null
  is_verified: boolean
  verified_at?: string | null
  contact_name?: string | null
  telegram_username?: string | null
  telegram_user_id?: number | null
  owner_telegram_user_id?: number | null
  owner_telegram_username?: string | null
  owner_display_name?: string | null
  image_path?: string | null
  image_filename?: string | null
  image_mime_type?: string | null
  created_at: string
  updated_at: string
  resolved_at?: string | null
  deleted_at?: string | null
}

export interface NewItemPayload {
  title: string
  description: string
  category: string
  location: string
  status: ItemStatus
  contact_name: string
  telegram_username?: string
  telegram_user_id?: number
  owner_telegram_user_id?: number
  owner_telegram_username?: string
  owner_display_name?: string
}


export interface MatchResult {
  id: number
  title: string
  status: ItemStatus
  category: string
  location: string
  relevance_score: number
  confidence: 'low' | 'medium' | 'high'
  reasons: string[]
  telegram_user_id?: number | null
  image_path?: string | null
}

export interface Claim {
  id: number
  source_item_id: number
  target_item_id: number
  requester_telegram_user_id?: number | null
  owner_telegram_user_id?: number | null
  status: 'pending' | 'approved' | 'rejected' | 'cancelled' | 'completed' | 'not_match'
  claim_message?: string | null
  source_item_title?: string | null
  target_item_title?: string | null
  shared_source_contact?: string | null
  shared_target_contact?: string | null
  shared_source_address?: string | null
  shared_target_address?: string | null
}
